### PR TITLE
[TOPIC GPIO] Adapt CC1200/CC2520 to new GPIO API

### PR DESCRIPTION
**Pull request by tbursztyka**

### CODE REVIEW

**Comment by tbursztyka** [comment]
@pabigot and doing the same here... pulling the 186 commits from the branch.

**Comment by carlescufi** [comment]
@tbursztyka can you please rebase against the latest `topic-gpio` branch?

**Comment by tbursztyka** [comment]
@carlescufi already did. github seems a bit lost, let me try something.

File: drivers/ieee802154/ieee802154_cc2520.c [hunk 1]
```diff
@@ -354,24 +354,24 @@ static inline void set_reset(struct device *dev, u32_t value)
...
+		     cc2520->gpios[CC2520_GPIO_IDX_VREG_EN].pin, value);
 }
 
 static inline u32_t get_fifo(struct cc2520_context *cc2520)
 {
 	u32_t pin_value;
 
-	gpio_pin_read(cc2520->gpios[CC2520_GPIO_IDX_FIFO].dev,
-		      cc2520->gpios[CC2520_GPIO_IDX_FIFO].pin, &pin_value);
+	pin_value = gpio_pin_get(cc2520->gpios[CC2520_GPIO_IDX_FIFO].dev,
```
Comment (pabigot) [inline drivers/ieee802154/ieee802154_cc2520.c]: Note that `gpio_pin_get` can return a negative value to indicate that the read attempt failed.  We've generally been returning `pin_value > 0` if the API isn't prepared to deal with runtime failures.

File: drivers/ieee802154/ieee802154_cc2520.c [hunk 2]
```diff
@@ -973,6 +963,84 @@ static int power_on_and_setup(struct device *dev)
...
+
+	/* VREG_EN */
+	gpio = device_get_binding(DT_CC2520_GPIO_VREG_EN_DRV_NAME);
+	if (!gpio) {
+		return NULL;
+	}
+
+	cc2520->gpios[CC2520_GPIO_IDX_VREG_EN].pin = DT_CC2520_GPIO_VREG_EN;
+	gpio_pin_configure(gpio, cc2520->gpios[CC2520_GPIO_IDX_VREG_EN].pin,
+			   flags_noint_out);
```
Comment (pabigot) [inline drivers/ieee802154/ieee802154_cc2520.c]: All pin configuration must incorporate the devicetree flags along with the direction flags.

File: dts/bindings/ieee802154/ti,cc1200.yaml [hunk 3]
```diff
@@ -6,3 +6,8 @@ description: Texas Instruments CC1200 802.15.4 wireless transceiver
 compatible: "ti,cc1200"
 
 include: spi-device.yaml
+
+properties:
+    int-gpios:
+      type: phandle-array
+      required: true
```
Comment (pabigot) [inline dts/bindings/ieee802154/ti,cc1200.yaml]: The standard practice now (see porting comment) is to document the active level for both input and output signals as perceived by the device.

**Review by pabigot — CHANGES_REQUESTED** [review]
Always add the flags from the devicetree binding when using `gpio_pin_configure()`.

File: drivers/ieee802154/ieee802154_cc2520.c (hunk 1 above)
Comment (tbursztyka) [reply drivers/ieee802154/ieee802154_cc2520.c]: That's a logic change in the driver. I am not going to change driver logic atm (no test setup etc...). So let's assume (as it was before) that it's always successful.

File: dts/bindings/ieee802154/ti,cc1200.yaml (hunk 3 above)
Comment (tbursztyka) [reply dts/bindings/ieee802154/ti,cc1200.yaml]: Could you pin-point to an actual example here?

File: drivers/ieee802154/ieee802154_cc2520.c (hunk 1 above)
Comment (pabigot) [reply drivers/ieee802154/ieee802154_cc2520.c]: It's not a change in logic, though it does fix a bug.

In the old API any error would have been in the return value which is discarded, and the pin_value would have been returned as an uninitialized value from the stack.

Here if you return `pin_value > 0` you will return zero in error and low, and 1 in active FIFO signal, which is robust and handles the error case that was previously ignored in a well-defined way.

**Comment by pabigot** [comment]
@tbursztyka Looking closely at the code, I see your point: the code uses literal 0 and 1 to control these lines, so devicetree flags may be wrong.

To accommodate that `gpio_pin_{set,get}_raw()` should be used everywhere `gpio_pin_{set,get}()` is currently used.  The devicetree flags should still be added to the configure calls; by using the correct raw API they will be ignored.

I will queue a patch to do this if it's not something you can get to.

**Comment by tbursztyka** [comment]
@pabigot you lost me: gpio_pin_{get, set}() make no use of any flags.

**Comment by pabigot** [comment]
Yes, they do.  They respect the configured active level.

**Comment by tbursztyka** [comment]
ok internally. I'll switch to raw then

**Comment by pabigot** [comment]
We'll need to discuss this.  The branch doesn't build for CC2520 because an include file that's still referenced was removed.

My understanding is the criteria for conversion of untestable hardware is that it build.  If it's just that CI passes (because it isn't even build-tested) then maybe this can go in.  Otherwise it needs more work.

**Comment by pabigot** [comment]
The devicetree names referenced in this PR are wrong.  Please run the following command with the following script:

    sed -i -r -f x.sed drivers/ieee802154/*.c

contents of x.sed:
```
s/DT_CC(....)_GPIO_(.*)_DRV_NAME/DT_INST_0_TI_CC\1_\2_GPIOS_CONTROLLER/g
s/DT_CC(....)_GPIO_(.*)_FLAGS/DT_INST_0_TI_CC\1_\2_GPIOS_FLAGS/g
s/DT_CC(....)_GPIO_(.*)\b/DT_INST_0_TI_CC\1_\2_GPIOS_PIN/g
```

Alternatively cherry-pick and squash https://github.com/pabigot/zephyr/commit/745c7896a5825d1b7654728734c860cd16b4ac2c which is in https://github.com/pabigot/zephyr/commits/tg/aug22241 which has a hack to allow this to be built for nrf52_pca10040.

Thanks.

**Comment by tbursztyka** [comment]
> The devicetree names referenced in this PR are wrong. 

It used to be that DT names of SPI slave devices would include the SPI master device DT bate name in it. It's not the case anymore?

**Comment by pabigot** [comment]
> > The devicetree names referenced in this PR are wrong.
> 
> It used to be that DT names of SPI slave devices would include the SPI master device DT bate name in it. It's not the case anymore?

There are several prefixes.  The ones used in this PR will not build.  The one I suggest is the shortest that does compile and is the standard format for drivers.

**Comment by pabigot** [comment]
Tagged DNM until devicetree names merged (I need a visible cue to help me distinguish this from 22156).

**Pull request closed without merging**